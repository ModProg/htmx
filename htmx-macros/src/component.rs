use attribute_derive::{FlagOrValue, FromAttr};
use manyhow::{bail, ensure, error_message, Result};
use quote::ToTokens;
use syn::{
    parse2, Attribute, Expr, FnArg, Ident, ItemFn, ItemStruct, Pat, PatIdent, PatTupleStruct,
    PatType, Signature, Type,
};

use crate::*;

pub fn component(_input: TokenStream, item: TokenStream) -> Result {
    let htmx = &htmx_crate();
    Ok(if let Ok(mut item) = parse2::<ItemStruct>(item.clone()) {
        // Generate
        // #[derive(typed_builder::TypedBuilder)]
        // #[builder(crate_module_path=htmx::__private::typed_builder)]
        // #[builder(build_method(into = Html))]

        // Set some defaults
        for field in item.fields.iter_mut() {
            type_attrs(
                field.ident.as_ref().ok_or_else(|| {
                    error_message!("Only structs with named fields are supported")
                })?,
                &field.ty,
                &mut field.attrs,
                htmx,
            )?;
        }
        quote! {
            #use #htmx::Html;
            #use #htmx::__private::typed_builder::{self, TypedBuilder};
            #[derive(TypedBuilder)]
            #[builder(crate_module_path = typed_builder)]
            #[builder(build_method(into = Html))]
            #item
        }
    } else if let Ok(ItemFn {
        attrs,
        vis,
        sig:
            Signature {
                ident,
                generics,
                inputs,
                output: _todo_assert_that_this_is_html,
                ..
            },
        block,
    }) = parse2::<ItemFn>(item.clone())
    {
        // #[derive(typed_builder::TypedBuilder)]
        // #[builder(crate_module_path=::typed_builder)]
        // #[builder(build_method(into = Html))]
        // struct MyFnComponent {
        //     a: bool,
        //     b: String,
        // }
        //
        // impl Into<Html> for MyFnComponent {
        //     fn into(self) -> Html {
        //         let Self { a, b } = self;
        //         htmx! {crate
        //             <button disabled=a> {b} </button>
        //         }
        //     }
        // }
        ensure!(generics.params.is_empty(), "generics are not supported");

        let (fields, fieldpats): (Vec<_>, Vec<_>) = inputs
            .into_iter()
            .map(|arg| {
                // hi
                ensure!(let FnArg::Typed(PatType { mut attrs, pat, ty, .. }) = arg,
                    arg, "`self` is not supported");

                let ident = match &*pat {
                    Pat::Ident(PatIdent { ident, .. }) => ident,
                    // On tuples with a single field, take its ident
                    Pat::TupleStruct(PatTupleStruct { elems, .. })
                        if elems.len() == 1 && matches!(elems.first().unwrap(), Pat::Ident(_)) =>
                    {
                        let Some(Pat::Ident(PatIdent { ident, .. })) = elems.first() else {
                            unreachable!("pat should contain a single ident")
                        };
                        ident
                    }
                    pat => bail!(pat, "only named arguments and new type patterns are allowed";
                    help = "use `ident @ {}`", pat.into_token_stream();),
                };

                type_attrs(ident, &ty, &mut attrs, htmx)?;

                Ok((quote!(#(#attrs)* pub #ident: #ty,), quote!(#ident: #pat,)))
            })
            .collect::<Result<Vec<_>>>()?
            .into_iter()
            .unzip();

        // #attrs #vis struct
        quote! {
            #use #htmx::Html;
            #use #htmx::__private::typed_builder::{self, TypedBuilder};
            #[derive(TypedBuilder)]
            #[builder(crate_module_path=typed_builder)]
            #[builder(build_method(into=Html))]
            #(#attrs)*
            #vis struct #ident {
                #(#fields)*
            }

            impl From<#ident> for Html {
                #[allow(non_shorthand_field_patterns)]
                fn from(#ident{ #(#fieldpats)* }: #ident) -> Html #block
            }
        }
    } else {
        bail!(item, "only functions and structs are supported")
    })
}
// todo derive macro
// #[component]
// fn MyFnComponent(a: bool, b: String) -> Html {
//     htmx! {crate
//         <button disabled=a> {b} </button>
//     }
// }
//
// // Generates
//
// #[derive(typed_builder::TypedBuilder)]
// #[builder(crate_module_path=::typed_builder)]
// #[builder(build_method(into = Html))]
// struct MyFnComponent {
//     a: bool,
//     b: String,
// }
//
// impl Into<Html> for MyFnComponent {
//     fn into(self) -> Html {
//         let Self { a, b } = self;
//         htmx! {crate
//             <button disabled=a> {b} </button>
//         }
//     }
// }
//
// // Using only struct
// #[derive(Component)]
// struct MyStructComponent {
//     a: bool,
//     b: String,
// }
// impl Into<Html> for MyStructComponent {
//     fn into(self) -> Html {
//         let Self { a, b } = self;
//         htmx! {crate
//             <button disabled=a> {b} </button>
//         }
//     }
// }
//
//

#[derive(FromAttr)]
#[attribute(ident = component)]
struct FieldAttribute {
    default: FlagOrValue<Expr>,
    children: bool,
}

fn type_attrs(
    name: &Ident,
    ty: &Type,
    attrs: &mut Vec<Attribute>,
    htmx: &TokenStream,
) -> Result<()> {
    let FieldAttribute { default, children } = FieldAttribute::remove_attributes(attrs)?;
    if let Type::Path(path) = ty {
        // TODO strip Option
        if default.is_flag() || path.path.is_ident("bool") || path.path.is_ident("Option") {
            attrs.push(parse_quote!(#[builder(default)]))
        }
        if let Some(default) = default.as_value() {
            attrs.push(parse_quote!(#[builder(default = #default)]))
        }
        if children
            || path.path.is_ident("Children")
            || path.path == parse_quote!(htmx::Children)
            || path.path == parse_quote!(#htmx::Children)
        {
            attrs.push(parse_quote! {#[builder(via_mutators, mutators(
                fn child(&mut self, $child: impl #htmx::ToHtml) {
                    self.#name.push($child);
                }
            ))]});
            attrs.push(parse_quote!(#[allow(missing_docs)]));
        }
    }
    attrs.push(parse_quote!(#[builder(setter(into))]));
    Ok(())
}
