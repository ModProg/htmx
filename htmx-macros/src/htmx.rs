use htmx_script::{Script, ToJs};
use manyhow::{ensure, Result};
use proc_macro2::{TokenStream, TokenTree};
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use crate::special_components::Node;
use crate::*;

pub fn htmx(input: TokenStream) -> Result {
    // https://github.com/rust-lang/rust-analyzer/issues/15572
    // let htmx = match (
    //     proc_macro_crate::crate_name("htmx"),
    //     std::env::var("CARGO_CRATE_NAME").as_deref(),
    // ) { (Ok(FoundCrate::Itself), Ok("htmx")) => quote!(crate),
    //   (Ok(FoundCrate::Name(name)), _) => { let ident = Ident::new(&name,
    //   Span::call_site()); quote!(::#ident) } _ => quote!(::htmx),
    // };
    let htmx = htmx_crate();

    let mut input = input.into_iter().peekable();

    let htmx = match input.peek() {
        Some(TokenTree::Ident(ident)) if ident == "crate" => {
            input.next();
            quote!(crate)
        }
        _ => htmx,
    };

    if input.peek().is_none() {
        return Ok(quote!(#htmx::Html::new()));
    }

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .raw_text_elements(["script"].into())
            .custom_node(),
    )
    // TODO parse_recoverable
    .parse_simple(input.collect::<TokenStream>())?
    .into_iter()
    .map(|n| expand_node(n, &htmx, false))
    .collect::<Result>()?;

    Ok(quote! {
        #use #htmx::{ToHtml, Html, IntoHtmlElements};
        {
            use #htmx::native::*;
            let mut $htmx = Html::new();
            #nodes
            $htmx
        }
    })
}

pub fn expand_node(node: Node, htmx: &TokenStream, child: bool) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name, htmx)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!($node = $node.child(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!($node = $node.child({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!($node = $node.child(#script);)
                }
            } else {
                expand_nodes(children, htmx, true)?
            };
            let main = quote!({let mut $node = #name #(.#attributes)*; #children $node.build()});

            let main = match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({let _: #name; #main})
                }
                _ => main,
            };
            if child {
                quote!($node = $node.child(&#main);)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&#main, &mut $htmx);)
            }
        }
        Node::Block(_) | Node::Text(_) => {
            if child {
                quote!($node = $node.child(&{#[allow(unused_braces)] #node});)
            } else {
                quote!(#htmx::ToHtml::write_to_html(&{#[allow(unused_braces)] #node}, &mut $htmx);)
            }
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node(htmx, child)?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>, htmx: &TokenStream, child: bool) -> Result {
    nodes
        .into_iter()
        .map(|n| expand_node(n, htmx, child))
        .collect()
}

fn name_to_struct(name: NodeName, htmx: &TokenStream) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::builder()), false)),
        name @ NodeName::Punctuated(_) => {
            let name = name.to_string();
            ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
         "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        );
            Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = name.value();
                ensure!(name.to_ascii_lowercase().chars().all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
                 "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
                );
                Ok((quote!(#htmx::CustomElement::new_unchecked(#name)), true))
            } else {
                Ok((quote!(#htmx::CustomElement::new(#name)), true))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
