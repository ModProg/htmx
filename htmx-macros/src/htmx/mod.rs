mod html;
mod special_components;

mod rusty;

pub use html::html;
pub use rusty::html as rtml;

// pub fn html(input: TokenStream) -> Result {
//     if input.is_empty() {
//         return Ok(quote!(::htmx::Html::new()));
//     }

//     let mut fork = input.clone().into_iter();

//     let first = fork.next();
//     let second = fork.next();

//     // TODO figure out actual differentiator
//     // probably would be, starts with `<` or starts with `{}` or `""` not
// followed     // by `,`

//     if matches!(input.peek(), Some(TokenTree::Punct(punct)) if
// punct.as_char() == '<') {         html::html(input.collect())
//     } else {
//         rusty::html(input.collect())
//     }
// }
