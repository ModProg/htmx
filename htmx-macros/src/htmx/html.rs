use htmx_script::{Script, ToJs};
use manyhow::{ensure, ErrorMessage, Result};
use proc_macro2::TokenStream;
use proc_macro_utils::TokenStream2Ext;
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use super::special_components::{Node, Special};
use crate::*;

pub fn html(input: TokenStream) -> Result {
    let mut parser = input.clone().parser();
    let (input, expr) =
        if let ( Some(expr), Some(_)) = (parser.next_expression(), parser.next_tt_fat_arrow()) {
            (parser.into_token_stream(), quote!(&mut #expr))
        } else {
            (input, quote!(::htmx::Html::new()))
        };

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .custom_node::<Special>()
            .raw_text_elements(["script"].into()),
    )
    // TODO parse_recoverable
    .parse_simple(input)?
    .into_iter()
    .map(expand_node)
    .collect::<Result>()?;

    Ok(quote! {
        {
            use ::htmx::native::*;
            let mut __html = #expr;
            #nodes
            __html
        }
    })
}

pub fn expand_node(node: Node) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                // TODO scripts
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!(__html.child_expr(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!(__html.child_expr({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!(__html.child(#script);)
                }
            } else {
                expand_nodes(children)?
            };
            let body = (!children.is_empty()).then(|| quote!(let mut __html = __html.body();));
            let main = quote!({let mut __html = #name #(__html.#attributes;)* #body; #children});

            match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({#name::unused(); #main})
                }
                _ => main,
            }
        }
        Node::Block(_) | Node::Text(_) => {
            quote!(::htmx::ToHtml::to_html(&{#[allow(unused_braces)] #node}, &mut __html);)
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node()?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>) -> Result {
    nodes.into_iter().map(expand_node).collect()
}

pub fn ensure_tag_name(name: String, span: impl ToTokens) -> Result<String, ErrorMessage> {
    ensure!(
        name.to_ascii_lowercase().chars()
            .all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
        span,
        "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        // TODO similar function but with css error: https://drafts.csswg.org/css-syntax-3/#non-ascii-ident-code-point
    );
    Ok(name)
}

fn name_to_struct(name: NodeName) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::new(&mut __html);), false)),
        name @ NodeName::Punctuated(_) => {
            let name = ensure_tag_name(name.to_string(), name)?;
            Ok((
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                true,
            ))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = ensure_tag_name(name.value(), name)?;
                Ok((
                    quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                    true,
                ))
            } else {
                Ok((
                    quote!(::htmx::CustomElement::new(&mut __html, #name);),
                    true,
                ))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
