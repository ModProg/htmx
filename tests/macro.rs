use htmx::{htmx, Html};
use htmx_macros::component;

struct Custom {
    href: String,
}

impl Custom {
    fn builder() -> Self {
        Custom {
            href: "default".to_owned(),
        }
    }

    fn href(mut self, value: impl Into<String>) -> Self {
        self.href = value.into();
        self
    }

    fn build(self) -> Html {
        htmx! {
            <a href=self.href/>
        }
    }
}

#[test]
fn test() {
    let html = htmx!(
        <div>
            <a href="hello" download/>
            <a href="hello" download="file.name"/>
            <Custom href="test"/>
        </div>
    );
    insta::assert_snapshot!(html.to_string());
}

#[test]
fn struct_component() {
    #[component]
    struct Component {
        a: bool,
        b: String,
    }

    impl From<Component> for Html {
        fn from(Component { a, b }: Component) -> Self {
            htmx! {
                <button disabled=a>{b}</button>
            }
        }
    }

    insta::assert_snapshot!(
        htmx! {
            <Component a b="Disabled Button"/>
            <Component a=true b="Disabled Button"/>
            <Component a=false b="Enabled Button"/>
            <Component b="Enabled Button"/>
        }
        .to_string()
    );
}

#[test]
fn fn_component() {
    #[component]
    fn Component(a: bool, b: String) -> Html {
        htmx! {
            <button disabled=a>{b}</button>
        }
    }

    insta::assert_snapshot!(
        htmx! {
            <Component a b="Disabled Button"/>
            <Component a=true b="Disabled Button"/>
            <Component a=false b="Enabled Button"/>
            <Component b="Enabled Button"/>
        }
        .to_string()
    );
}

#[test]
fn reserved_attributes() {
    insta::assert_snapshot!(
        htmx! {
            <script type_="module" />
            <script async_=true />
        }
        .to_string()
    );
}

#[test]
fn custom_element() {
    insta::assert_snapshot!(
        htmx! {
            <custom-element attr="module">
                <p> "This is a child" </p>
            </_>
            <{"div"} custom_div="hello"> </_>
        }
        .to_string()
    );
}
