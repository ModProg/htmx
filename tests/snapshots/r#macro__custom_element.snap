---
source: tests/macro.rs
expression: "htmx! {\n            < custom - element attr = \"module\" > < p > \"This is a child\" < / p\n            > < / _ > < { \"div\" } custom_div = \"hello\" > < / _ >\n        }.to_string()"
---
<!DOCTYPE html><custom-element attr="module"><p>This is a child</p></custom-element><div custom_div="hello"></div>
